Refactor Xiaomi vacuum to sensors

OGKevin opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

Attributes from the vacuum platform have been replaced with sensors.
This means that all your references to these attributes must be replaced with its sensor.
Some sensors have been disabled by default and need manual enablement once the integration has been set up.

e.g. the attribute `mop_attached: true` has now been replaced with `binary_sensor.<NAME>_mop_attached`

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #51361
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19641

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

OGKevin committed:
Refactor Xiaomi vacuum with sensors.

This is the first step into refactoring xiaomi vacuum attributes into sensors.
What is missing are some switches and binary sensors etc.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Use generic sensor for Xiaomi vacuum sensors.

By using HA coordinator, the generic Xiaomi sensor class can be used with these
coordinators to get the status sensors from vacuum. This also means now that
sensors are available as soon as HA starts, which is a nice plus.

Now the only reason to create a subclass of the generic sensors is when custom
value parsing needs to be done.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Working vacuum sensors via 1 coordinator.

Vacuum needs a custom coordinator to ensure that it fetches all the needed data
and puts it in a dict. From this dict the sensors will then get their data
accordingly.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Remove vacuum setup method in sensor

Sensor is generic enough that vacuum does not require its own setup method.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Don't auto register generic sensors.

Let the user decide which sensor is useful for them and enable them.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

rytilahti reviewed (commented):
Great work @OGKevin! I tested this briefly using a gen1 vacuum; the sensors are going unavailable very frequently, the reason is most likely caused by concurrent data requests causing the vacuum to ignore "passed" requests.

![image](https://user-images.githubusercontent.com/3705853/130335891-65884b71-e73b-485a-9e8d-88f0fcf9a7d6.png)

This doesn't seem to affect to the vacuum entities state though, which is odd:
![image](https://user-images.githubusercontent.com/3705853/130335906-ccd866ed-468c-4978-83d0-cd91ad9e2167.png)


Open questions:
* Icons
* Durations in minutes instead of seconds?
* Naming of sensors - should they include the parent device name at all? Should these be made somehow translatable?

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -108,6 +115,48 @@ def get_platforms(config_entry):
+        """Fetch data from the device using async_add_executor_job."""
+        try:
+            async with async_timeout.timeout(10):
+                results = await asyncio.gather(
+                    hass.async_add_executor_job(device.status),
+                    hass.async_add_executor_job(device.dnd_status),
+                    hass.async_add_executor_job(device.last_clean_details),
+                    hass.async_add_executor_job(device.consumable_status),
+                    hass.async_add_executor_job(device.clean_history),
+                )
```
Yes, we should create a function that does all these sync calls and then schedule that function once on the executor here.

OGKevin replied:
Made them sequential. @rytilahti when you have a chance, pls let me know if it solved the issue for you. Looks good on my end.

rytilahti replied:
It is indeed much better! What should be changed is to modify the vacuum platform to use the data from the coordinator instead of doing separate polling (which causes currently failures now and then).

OGKevin replied:
aah good suggestion, will look into this as well then.

OGKevin commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -232,6 +260,106 @@ class XiaomiMiioSensorDescription(SensorEntityDescription):
         entity_registry_enabled_default=False,
     ),
+    ATTR_DND_START: XiaomiMiioSensorDescription(
+        key=ATTR_DND_START.split(ATTR_DND)[1],
```
If we move to 1 container per platform then the .split is not needed. As we can store the key in the constant without overlaps. The reason split is needed is bc we only need the word `start` as key/attribute. Instead of `do_not_disturb_start`

OGKevin commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -232,6 +260,106 @@ class XiaomiMiioSensorDescription(SensorEntityDescription):
+        name="Last clean area",
+    ),
+    ATTR_CLEAN_HISTORY_TOTAL_DURATION: XiaomiMiioSensorDescription(
+        unit_of_measurement=TIME_SECONDS,
```
Yea that is what I was thinking as well, provide the precisions from the device, and then users can use template sensors maybe to convert it to something they want? 

Seeing that all the sensors are disabled by default, we can also add multiple, 1 in seconds, minutes, and hours etc.

MartinHjelmare replied:
For now, just provide the data as is in one disabled by default sensor.

OGKevin commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -421,19 +568,36 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
         self._attr_name = name
         self._attr_unique_id = unique_id
-        self.entity_description = description
+        self._attr_entity_registry_enabled_default = False
```
I agree and disagree.

Because this is a `XiaomiGenericSensor` it should be disabled by default. We can either:
1. extend classes from this one and set it to true
2. make it configurable via the description

Registring all generic sensors blindly is not nice and fills user's instances with a lot of sensors that they might never use.

rytilahti replied:
Agreed, we should parametrize it for the generic sensor / description, and make it default to `False`. I'd still like to have things like "last clean time" and "last clean area" enabled per default, as that information is likely to be used by most of the people.

OGKevin commented on homeassistant/components/xiaomi_miio/const.py:
```diff
@@ -135,7 +140,7 @@
     + MODELS_LIGHT_MONO
 )
-MODELS_VACUUM = ["roborock.vacuum", "rockrobo.vacuum"]
+MODELS_VACUUM = ["roborock.vacuum", "rockrobo.vacuum", "roborock.vacuum.a15"]
```
Seeing that there is also interest in limiting sensors per device etc, it would be better to then list all the models.

A lot of checks also have the `model in MODELS_VACUUM` rather than prefix test.

OGKevin replied:
I'll make the needed modifications and sensor registrations appropriately.

OGKevin commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -232,6 +260,106 @@ class XiaomiMiioSensorDescription(SensorEntityDescription):
         state_class=STATE_CLASS_TOTAL_INCREASING,
         entity_registry_enabled_default=False,
     ),
+    ATTR_DND_START: XiaomiMiioSensorDescription(
```
yea, separate container will be better indeed.

OGKevin committed:
Remove converted attributes from xiaomi vacuum.

The attributes that have been converted so far should be removed from the vacuum
attributes list.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Fetch data from vacuum sequentially.

It seems some vacuums do not like parallel requests. The requests that came
before are ignored.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Refactor vacuum sensors to its own container.

By moving vacuum sensors to its own container, there is no more key collisions.
This in turns means that there is need for the split hack to ensure key names
are correct.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
fixup! fix bad rebase.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Fix sensor naming and default registration.

Use proper names for sensors, no need to include from which device status it
came.

Registration of the sensor by default has been parameterised. If the param is
not set, the sensor is not registered.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Make vacuum platform also use coordinator for its data.

By using the coordinator for data in vacuum platfrom, removes the cases where
request gets ignored during the case where the requests are done concurrently by
separate platforms.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Add binary sensor for vacuum

Add binary sensor for waterbox, mop, and water shortage.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Added proper icons to sensors.

https://github.com/home-assistant/core/issues/51361

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

rytilahti commented:
Works much better now, as the device doesn't get confused by two sets of sequence ids! :+1: I haven't tested any actions on the vacuums, just that the data gets updated.

The water shortage sensors is telling me "Ok", but I didn't yet dig into why that's happening. box & mop attached are showing "Disconnected", which I suppose is fine for now even though the vacuum has none of those -- when we get a better view of which specific features are supported by model, we can adapt that if even needed.

rytilahti commented on homeassistant/components/xiaomi_miio/vacuum.py:
```diff
@@ -285,43 +295,11 @@ def extra_state_attributes(self):
-                    ),
                     ATTR_STATUS: str(self.vacuum_state.state),
                     ATTR_MOP_ATTACHED: self.vacuum_state.is_water_box_attached,
                 }
```
The mop attached is already a sensor, so that could be removed. Should we expose the error & the state (which is the more fine-grained variant what's the state homeassistant uses as a main state) as sensors, too?

OGKevin replied:
yea i was planning indeed to make a switch sensor for error and state

OGKevin replied:
But not 100 sure if that is ok, my idea to make it a select sensor is bc you then can use this in automations etc pretty nicely. But then ofc the actual option selector will not work.

But the fact that the state is already translated, do we really need to add it?

MartinHjelmare replied:
I suggest not exposing the error as any additional sensors for now. Just follow the entity design for now.

https://developers.home-assistant.io/docs/core/entity/vacuum

Extra state attributes not part of the entity model should be avoided though. Those are better exposed as separate (binary) sensors.

OGKevin replied:
Yea based on this, then i do not think status and error should be a sep sensor 🤔. Unless we explicit want to expose a more detailed state via sensor that is specific to xiaomi vacuums. But then imo we should have use cases. e.g if DOCKER != CHARGING, then we can expose a binary charging sensor etc etc.

rytilahti replied:
I agree that there's no much reason to expose the status separately (l thought there was a different mode for mopping which might have been useful for someone, but alas, there isn't). I can see some use for the error state (inform about which error happened), but that can be added later if wanted.

rytilahti commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -108,6 +115,58 @@ def get_platforms(config_entry):
+            KEY_VACUUM_CONSUMABLE_STATUS: device.consumable_status(),
+            KEY_VACUUM_CLEAN_HISTORY_STATUS: device.clean_history(),
+        }
+
```
I don't think it's "homeassistant-esque", but if it's allowed by the coding standards / reviewers, I would rather get rid of the overuse of consts inside the integration, and rather use a dataclass-based container for the coordinator data.

This would allow easier extending later on thanks to type hints:
```
@dataclass
class VacuumCoordinatorData:
    status: VacuumStatus
    dnd_status: VacuumDnDStatus
    ...
```

rytilahti commented on homeassistant/components/xiaomi_miio/binary_sensor.py:
```diff
@@ -110,5 +182,11 @@ def _extract_value_from_attribute(state, attribute):
+            return value
+
+        _LOGGER.warning(
+            f"could not determine how to parse state value of type: {type(value)}"
```
```suggestion
            "could not determine how to parse state value of type: %s", type(value)
```

rytilahti commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -312,6 +336,142 @@ class XiaomiMiioSensorDescription(SensorEntityDescription):
+        name="Last Clean Start",
+        device_class=DEVICE_CLASS_TIMESTAMP,
+        parent_key=KEY_VACUUM_LAST_CLEAN_STATUS,
+        entity_registry_enabled_default=True,
```
```suggestion
```
`True` is the default value for this in `EntityDescription`, so I think it's worth only to explicitly mark those with `entity_registry_enabled_default=False` that are not to be enabled per default.

OGKevin commented:
@rytilahti is `None` an expected value? Or does this mean something went/is wrong?

OGKevin committed:
Refactor sensors to use dataclass.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

rytilahti commented:
> @rytilahti is `None` an expected value? Or does this mean something went/is wrong?

None means currently that the device didn't deliver a response for that, i.e., it's unsupported. Looks like this is not consistent among `is_water_*` (https://github.com/rytilahti/python-miio/blob/master/miio/vacuumcontainers.py#L189)..

OGKevin commented:
Alrighty then, so what is left is:

* figure out why I can't run test locally
* Fix failing tests
* Add (water/mop) sensors based on models

MartinHjelmare commented on homeassistant/components/xiaomi_miio/vacuum.py:
```diff
@@ -209,22 +218,22 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
         """Initialize the Xiaomi vacuum cleaner robot handler."""
         super().__init__(name, device, entry, unique_id)
 
         self.vacuum_state = None
```
Consider removing this attribute and just access the coordinator.data instead in the properties.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/vacuum.py:
```diff
@@ -121,7 +124,13 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
         _LOGGER.debug("Initializing with host %s (token %s...)", host, token[:5])
         vacuum = Vacuum(host, token)
```
We should probably not create a new device here but use the same device as the coordinator is using.

OGKevin committed:
Make vacuum use coordinator for its data.

This commit also insures that the binary sensors are only registered for devices
that include a mop.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin commented on homeassistant/components/xiaomi_miio/vacuum.py:
```diff
@@ -209,22 +218,22 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+
     @property
     def state(self):
         """Return the status of the vacuum cleaner."""
```
Where do you propose to move this particular logline?

Are you suggesting moving this logic to the coordinator?

MartinHjelmare replied:
If we want to keep the logging we can define `_handle_coordinator_update` in the entity class and set instance attributes with the state, and log if needed.

https://github.com/home-assistant/core/blob/0dc8fb497b1162c08286dfd03f54da3d237f38dd/homeassistant/helpers/update_coordinator.py#L322-L325

OGKevin replied:
This method is not available in the coordinator class that we are using here. The coordinator that is being used is this one: https://github.com/home-assistant/core/blob/f1a4ba8bb0a7b241e69eacc75f194a78d0490535/homeassistant/helpers/update_coordinator.py#L33

MartinHjelmare replied:
The coordinator will tell the coordinator entity to update state and the default method that handles the update in the CoordinatorEntity is  `_handle_coordinator_update`.

OGKevin replied:
aaah i get it now 👍🏾

OGKevin committed:
Merge remote-tracking branch 'origin/dev' into 51361-refactor-xiaomi-vacuum-to-sensors--from-dev

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin commented:
I'm still struggling to get the tests to work on my local env. 
Basically, I'm not able to test with my debugger on MacOS using pycharm.
This is slowing down fixing the broken tests and potentially rewriting them.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -1,6 +1,8 @@
 """Support for Xiaomi Miio."""
+from dataclasses import dataclass
```
```suggestion
from __future__ import annotations

from dataclasses import dataclass
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -62,7 +71,7 @@
 ]
 LIGHT_PLATFORMS = ["light"]
-VACUUM_PLATFORMS = ["vacuum"]
+VACUUM_PLATFORMS = ["vacuum", "sensor", "binary_sensor"]
```
Please sort 🔡.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -114,6 +123,97 @@ def get_platforms(config_entry):
+    last_clean_details: CleaningDetails
+    consumable_status: ConsumableStatus
+    clean_history_status: CleaningSummary
+    timers: List[Timer]
```
```suggestion
    timers: list[Timer]
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -114,6 +123,97 @@ def get_platforms(config_entry):
+    consumable_status: ConsumableStatus
+    clean_history_status: CleaningSummary
+    timers: List[Timer]
+    fan_speeds: Dict[str, int]
```
```suggestion
    fan_speeds: dict[str, int]
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -114,6 +123,97 @@ def get_platforms(config_entry):
+    clean_history_status: CleaningSummary
+    timers: List[Timer]
+    fan_speeds: Dict[str, int]
+    fan_speeds_reverse: Dict[int, str]
```
```suggestion
    fan_speeds_reverse: dict[int, str]
```

MartinHjelmare commented on homeassistant/components/xiaomi_miio/__init__.py:
```diff
@@ -114,6 +123,97 @@ def get_platforms(config_entry):
+    A class that holds attribute names for VacuumCoordinatorData.
+
+    These attributes can be used in methods like `getattr` when a generic solutions is needed.
+    See homeassistant.components.xiaomi_miio.device.XiaomiCoordinatedMiioEntity._extract_value_from_attribute for
```
Please break longs strings around 88 characters. Docstring lines should preferably not be longer than 72 characters according to PEP8.

OGKevin replied:
Linter did not complain tho :(

OGKevin replied:
Fixed my editor settings 👍🏾

MartinHjelmare commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -426,19 +579,42 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
         """Initialize the entity."""
         super().__init__(name, device, entry, unique_id, coordinator)
-
+        self._attr_device_class = description.device_class
```
We don't need to set these entity attributes if we set the entity description attribute, and the description has these corresponding attributes.

OGKevin committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

OGKevin committed:
Minor refactoring

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Update data from coordinator after running command.

This is is to have a faster status change when executing a command like changing
fan speeds. If a manual refresh is not triggered. Worst case scenario it will
take 10s for the new fan speed to be reported in HA.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

MartinHjelmare commented on homeassistant/components/xiaomi_miio/vacuum.py:
```diff
@@ -300,6 +300,7 @@ async def _try_command(self, mask_error, func, *args, **kwargs):
         """Call a vacuum command handling error messages."""
         try:
             await self.hass.async_add_executor_job(partial(func, *args, **kwargs))
+            await self.hass.async_run_job(self.coordinator.async_refresh)
```
Is there a reason to use `hass.async_run_job`? Normally we just await coroutines from inside coroutine functions.

OGKevin committed:
Refresh within coroutine is ok.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin commented:
@rytilahti is there some way you or another maintainer could have a look at the tests? I kind of give up.

OGKevin committed:
Merge remote-tracking branch 'origin/dev' into 51361-refactor-xiaomi-vacuum-to-sensors--from-dev

OGKevin committed:
Move logging to _handle_coordinator_update

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Use internal state attribute.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin commented:
I think this is ready except for the broken tests. 
I can't seem to get PyCharm configured to be able to run them properly with the debugger for this project.
Looking on the dev forums seems I'm not the only one.
This makes writing tests/code in python harder for me as it's such a "magical" language :).

OGKevin commented:
How shall we proceed?

OGKevin committed:
Merge remote-tracking branch 'origin/dev' into 51361-refactor-xiaomi-vacuum-to-sensors--from-dev

MartinHjelmare committed:
Fix vacuum typing

MartinHjelmare committed:
Fix tests constants

MartinHjelmare committed:
Fix vacuum inconsistent return values

MartinHjelmare committed:
Fix vacuum state update

MartinHjelmare committed:
Fix vacuum tests

MartinHjelmare committed:
Remove impossible cases

MartinHjelmare committed:
Parametrize custom services test

MartinHjelmare committed:
Move goto test

MartinHjelmare committed:
Move clean segment test

MartinHjelmare committed:
Move clean single segment test

MartinHjelmare committed:
Test service pause

MartinHjelmare committed:
Include vacuum in coverage

MartinHjelmare committed:
Delint

MartinHjelmare commented:
I've fixed parts of the code and the tests. Please do a manual test now again of the device.

OGKevin commented:
@MartinHjelmare thx btw, for picking up the tests 🤗 ❤️

rytilahti commented:
Sorry @OGKevin, I have been busy and I couldn't have helped much with the tests as we discussed previously. Thank you @MartinHjelmare for lending a hand here :1st_place_medal: 

I tested this with gen1 vacuum using manual configuration. The basic controls from the UI are working and sensors gets updated as expected after cleaning. Looks like mine is timeouting on every update cycle (python-miio retries automatically) making the updates take up to 5 seconds, are you seeing this behavior too, @OGKevin?

One quick question, should the `vacuum` platform also be listed on the device info page?

![image](https://user-images.githubusercontent.com/3705853/136224478-d148cf30-8742-4cb8-8188-a428b94e0c07.png)

![image](https://user-images.githubusercontent.com/3705853/136224571-8d383e39-83da-40c0-a829-b2a0dbaf3156.png)

MartinHjelmare commented:
Yes, the vacuum entity inherits the xiaomi miio coordinator entity which defines device info property.

bieniu commented:
I don't have a vacuum but I'll try to test the impact of this change on other platforms tonight.

rytilahti commented:
Ah, I think that's caused by the missing mac address information on my setup, which causes the missing connection? But that's not a blocker as it concerns only gen1 users without cloud connectivity. The upstream lib could be adjusted to return a dummy mac to make it work in the future, I suppose.

OGKevin commented:
hmmm
As I was working on the docs, I've noticed that the DnD sensors are missing.

```
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new binary_sensor.xiaomi_miio entity: binary_sensor.roborock_vacuum_a15_mop_attached
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new binary_sensor.xiaomi_miio entity: binary_sensor.roborock_vacuum_a15_water_box_attached
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new binary_sensor.xiaomi_miio entity: binary_sensor.roborock_vacuum_a15_water_shortage
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_last_clean_start
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_last_clean_end
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_last_clean_duration
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_last_clean_area
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_total_duration
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_total_clean_area
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_total_clean_count
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_total_dust_collection_count
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_main_brush_left
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_side_brush_left
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_filter_left
2021-10-07 12:22:32 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new sensor.xiaomi_miio entity: sensor.roborock_vacuum_a15_sensor_dirty_left
```

https://github.com/OGKevin/core/blob/d8489275fc3b602b780a06e14e258d4f019884e8/homeassistant/components/xiaomi_miio/sensor.py#L371-L386

Having a look at why this is 🤔.

OGKevin committed:
Fix vacuum sensor dict collision.

This also prevents collision for unique id. As the key is used to construct
unique ids.

OGKevin commented:
@MartinHjelmare added link to docs change.

OGKevin committed:
Use f strings as dict keys

rytilahti reviewed (approved):
Let's merge it then. Thank you @OGKevin and all reviewers for pushing this forward! :-)
